File: printcostcalculation.py
""" Code to read print orders from sample.csv file and output printing costs
    Input: CSV Data for Print jobs from .csv file
    Output: Printing cost for individual print jobs and total cost of printing
"""

#import module to open and read CSV files.
import csv

#create dictionary for print costs. Can be expanded to include other cost types.
a4PrintCosts = {'oneSidedBWCost': 0.15, 'oneSidedColorCost': 0.25, 'twoSidedBWCost': 0.10, 'twoSidedColorCost': 0.20}
currJobNum = 1
csvHasHeader = True #variable indicating that CSV file has a header. Can be switched for csv files without headers.
totalPrintCost = 0.0 #variable to hold total cost for printing

def csv_reader (file):	# function to read file object and return raw data
    return csv.reader(file)

#function for calculation of printing costs.
def calculatePrintJobCost(printData, costDict):
    bwPages = int(printData[0]) - int(printData[1])
    colorPages = int(printData[1])
    isTwoSided = printData[2].replace(' ', '')

    if isTwoSided == 'true':
        printJobCost = (bwPages * costDict.get('twoSidedBWCost')) + (colorPages * costDict.get('twoSidedColorCost'))
        return "{:.2f}".format(printJobCost)
    elif isTwoSided == 'false':
        printJobCost = (bwPages * costDict.get('oneSidedBWCost')) + (colorPages * costDict.get('oneSidedColorCost'))
        return "{:.2f}".format(printJobCost)

#main
if __name__ == "__main__":

    csvFilePath = "sample.csv" #assumes that csv file is in same folder as python file

    with open(csvFilePath, "r") as fileObject:
        reader = csv_reader(fileObject)

        for row in reader:	#looping through csv data for individual records
            if csvHasHeader:	#skips first line for csv files with headers
                csvHasHeader = False
            else:
                currentJobCost = calculatePrintJobCost(row, a4PrintCosts)
                print('Cost for print job ' + str(currJobNum) + ' is: $' + currentJobCost)
                currJobNum += 1
                totalPrintCost += float(currentJobCost)	#Outputs individual printing costs to console

        print('Total Printing Cost is: $' + "{:.2f}".format(totalPrintCost))	#Outputs total printing cost to console

File: printcostcalculation.go
//declaring this to be the main package
package main

//importing several necessary packages
import (
	"encoding/csv"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"math"
	"strconv"
	"strings"
)

//defining printing costs as constant values
const (
	oneSidedBWCost    float64 = 0.15
	oneSidedColorCost float64 = 0.25
	twoSidedBWCost    float64 = 0.10
	twoSidedColorCost float64 = 0.20
	//csvHasHeader      bool    = true
)

//error checking function
func check(e error) {
	if e != nil {
		panic(e)
	}
}

func round(num float64) int {
	return int(num + math.Copysign(0.55, num))
}

func toFixed(num float64, precision int) float64 {
	output := math.Pow(10, float64(precision))
	return float64(round(num*output)) / output
}

func main() {
	csvHasHeader := true
	totalPrintJobCost := 0.0
	dat, err := ioutil.ReadFile("sample.csv")
	check(err)

	r := csv.NewReader(strings.NewReader(string(dat)))

	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}

		//logic to calculate and output print job costs
		if csvHasHeader == true {
			csvHasHeader = false
		} else {
			totalPages, err := strconv.ParseFloat(record[0], 32)
			if err != nil {
				log.Fatal(err)
			}
			colorPages, err := strconv.ParseFloat(record[1], 32)
			if err != nil {
				log.Fatal(err)
			}
			bwPages := (totalPages - colorPages)
			isTwoSided, err := strconv.ParseBool(record[2])

			if isTwoSided == true {
				currentPrintJobCost := (bwPages * twoSidedBWCost)
				currentPrintJobCost += (colorPages * twoSidedColorCost)
				fmt.Printf("Cost for current print job is: $%.2f \n", currentPrintJobCost)
				totalPrintJobCost += currentPrintJobCost
			} else if isTwoSided == false {
				currentPrintJobCost := (bwPages * oneSidedBWCost)
				currentPrintJobCost += (colorPages * oneSidedColorCost)
				fmt.Printf("Cost for current print job is: $%.2f \n", currentPrintJobCost)
				totalPrintJobCost += currentPrintJobCost
			}
		}
	}
	fmt.Printf("Total Print Cost is: $%.2f", totalPrintJobCost)
}
